---
title: "Getting started with Phyloseq"
output:
  html_notebook: default
  pdf_document: default
---
ignore unless need to make pdf
```{r}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=60),tidy=TRUE)
```

Code written by Samara Zinman (Samara.zinman@gmail.com) and Edited by Sara Williams (swilliams@mote.org).

## What is phyloseq?

phyloseq is a R package for exploring microbiome profiles. The phyloseq package is a tool to import, store, analyze, and graph sequencing data that has already been clustered into OTUs or ASVs. This package uses many of the tools already available in R for ecology and phylogenetic analysis (packages like vegan ) and supports ggplot2 to easily make publication-ready figures.

A lot of great tutorials out there for Phyloseq, e.g. https://joey711.github.io/phyloseq/index.html

Link to R documentation and function list: https://www.rdocumentation.org/packages/phyloseq/versions/1.16.2 

Packages you will need for doing basic microbial analyses:
```{r Loading packages and set working directory}
#set your working directory OR the nice thing about R notebooks is that your wd is automatically the folder where your code is located.
#setwd("~/Mote_Coral_Fall2020/microbes_pipeline")
library(phyloseq)
library(ggplot2)
library(vegan)
#... add any other standard packages.
```

## Microbial Data
### File structures
- All of the data being loaded into R for this notebook, should be csv files. 
- Your sample names should be the same across all files
- Your ASV/OTU names must also be consistent across files

### Explanation of practice dataset we will use and getting it into R
This dataset is from a study of *Acropora cervicornis* microbiomes from fragments located in the same nursery but collected from (and previously located in nurseries) reefs across the Upper, Middle, and Lower Keys. This dataset contains an abundance, or OTU table, a taxonomy table, and a table of the sample data, AKA the meta data.
```{r Loading required csv files}
otudf <- read.csv("acer_otu_table.csv",row.names = 1) #loads csv file into R as a dataframe
#row.names = 1 is important for when this data is used later, because it sets the table's row names to the sample ID or taxa if taxa are rows so that it can be come a numeric matrix required later for otu_table()
taxdf <- read.csv("acer_tax_table.csv", row.names = 1) #loads csv file into R as a dataframe
#row.names = 1 is important for when this data is used later, because it sets the table's row names to the ASV, otherwise, it gets confused and sets the ID to the kingdom row
samdf <- read.csv("acer_sample_data.csv",row.names = 1) #loads csv file into R as a dataframe
#row.names = 1 is important for when this data is used later, because without it, r will not be able to match the sample names from the sample data to the otu table when building the phyloseq object
samdf
```

### What are the components of a phyloseq object?

- otu_table: Table of the abundance data, where sample IDs are rows and taxa (ASVs or OTUs) are columns or vice versa. Whether or not the taxa are rows depends on the analysis being done on the object. Be sure to set taxa_are_rows equal to TRUE if your taxa are rows.

- tax_table: Table of the taxonomy data, so the rows are ASVs or OTUs, and the columns are taxonomic rank (i.e. Kingdom, phylum, etc.)

- sample_data: Table of the metadata, explaining more information about your samples. This data MUST include sample IDs as the first row, and it MUST contain more than 2 columns or it will not compile.

```{r Make the components of the phyloseq object}
#################OTU_Table#########################
otumat<- as.matrix(otudf) #otu_table() doesn't like dataframes so you need to convert to a matrix
otu <- otu_table(otumat, taxa_are_rows = FALSE) #makes data into an otu table object
#be sure to confirm whether or not your taxa are rows or columns, in this case they are rows
#easy way to check this:
otumat[1:2,1:2] # the columns are ASVs

#################Tax_Table#########################
taxmat <- as.matrix(taxdf) #tax_table needs matrices not dataframes 
tax <- tax_table(taxmat) #makes data into a taxonomy table object

#################Sample_Data########################
sample.data <- sample_data(samdf) #makes a sample data table
#note: sample data does NOT need to be a matrix to compile

head(sample.data)
```

### Making a Phyloseq object
##### what goes in:
  When making a phyloseq object, there are 2 components necessary to getting the phyloseq object to   compile. These are your taxonomy table and your otu table (see the last chunk for how to create these components). However, in order to really do anything with your phyloseq object, you want to include a sample data table (see the last chunk for how to create this).        
  IMPORTANT NOTE: In order for these to compile, be sure your sample IDs and OTUs or ASVs match between tables

##### what comes out:
  When the phyloseq object is made, you get one object containing all of your abundance, sample, and taxonomy data. Those data are all linked so the abundance data and sample data overlap by sampleID, and the abundance data and taxonomy data overlap by OTUs or ASVs. This means that with the sampleID you can get data from 2 tables, and with the OTUs or ASVs you can get data from 2 tables. Phyloseq objects allow you to organize your data into 1 object containing all of your data, so if you make edits to your data, you only have to load it in the beginning. 

```{r Making the phyloseq object}
ps <- phyloseq(otu, tax, sample.data) #make the phyloseq object from the components made in the last chunk
ps #view the phyloseq object
#this object should have 1195 taxa and 64 samples
```

### Exploring your Phyloseq object
There are ways to view each component or your phyloseq object. In the below chunk, we use head() as a way to just look at the first 6 rows. Remove head() from the line to see then entire object.
```{r Exploring the phyloseq object}
#to view the sample data
head(sample_data(ps))

#to view the taxonomy table
head(data.frame(tax_table(ps))) 

#to view the otu table
#data.frame(otu_table(ps)) #this is how you actually view the otu table data, uncomment it
```

### Filtering, subsetting, and pruning your phyloseq object
#### Filtering and pruning
Sometimes, your phyloseq object is messy, from having a lot of taxa with little to no abundance at all included, for example from using raw abundances, or from having too many samples for what the tests or graphs you are running need, etc. There are many things you can do to clean your phyloseq object. This includes filtering it to exclude super small abundances, converting your raw abundance data to relative abundance data, or by merging samples.

```{r Filtering and pruning}

#get rid of 0 taxa
ps_clean <- prune_taxa(taxa_sums(ps) > 0, ps)
any(sample_sums(ps_clean) == 0) #reports FALSE if no zero taxa

#Remove Samples with less than 100 reads
ps_clean = prune_samples(sample_sums(ps_clean) > 100, ps_clean)

#You can filter taxa to remove taxa with small abundances, i.e. get rid of taxa that have less than 10 reads in 10% of the samples
ps_clean = filter_taxa(ps_clean, function(x) sum(x > 10) > (0.01*length(x)), TRUE) 

#Or alternatively, get rid of taxa in less than 4 samples
ps_clean<- prune_taxa(taxa_sums(ps_clean) > 4, ps_clean)

#get rid of common contaminants
ps.filt<-subset_taxa(ps_clean, (Order!="Chloroplast") | is.na(Order)) 
ps.filt<-subset_taxa(ps_clean, (Family!="Mitochondria") | is.na(Family)) 

#You can also convert your data from raw abundance data to relative abundance data
physeq.f.ra <- transform_sample_counts(ps.filt, function(x) x*100/sum(x)) #transform to relative abundance from raw abundance

rowSums(data.frame(otu_table(physeq.f.ra)))#if the line above worked right, all of the row sums should be 100

#You can also merge samples by a variable in the sample data, i.e. Region
ps.reg.merge <- merge_samples(ps,"Region")#the second part is the variable name be sure to spell it correctly, capitalize correctly, and keep it in quotations, if it doesn't match its name in your sample data, then this line WONT work

#Do not let the warning message from merge_samples() alarm you, it is not harmful
head(data.frame(otu_table(ps.reg.merge))) #if done right, this data is now combined so there is 1 abundance row in the otu table for each region

```
##### subsetting
Using the subset_samples() or subset_taxa() functions, you can subset your phyloseq object into a smaller part of the data, i.e. only the data from the upper keys. In order to do this, you need to know the different boolean operators in R.

Complete list of R operators here: https://www.statmethods.net/management/operators.html

```{r Subsetting your data}
#subsetting your data by info from the sample data, using ==
ps.up <- subset_samples(ps, Region=="Upper Keys") 
#this subsets your data by region equals Upper Keys, so you only get Upper Keys
ps.up #this object should have 1195 taxa and 13 samples

####subsetting you data by info from the sample data, using other boolean operators##########
ps.nolow <- subset_samples(ps, Region!="Lower Keys")
#this uses the != operator (AKA not equal to) to subset for everything except the lower keys
ps.nolow

#you can subset by more than just equal to or not equal to
ps.latlong <- subset_samples(ps, Lat>=24.5 && Long<=81.5) 
#this uses the && operator (AKA and) to subset by Lat and Long, this also uses the >= (AKA greater than or equal to) to subset Lat so it only includes latitudes >= 24.5, and it uses <= (AKA less that or equal to) to subset Long so it only includes longitudes <= 81.5
ps.latlong

#you can also subset by or not just and
ps.uplow <- subset_samples(ps, Region=="Upper Keys"|Region=="Lower Keys")
#this uses the | operator (AKA or) to subset, choosing any samples that are upper keys OR lower keys 
#this is different than && because && requires both halves to be true, but for | one half or the other can be true not necessarily both
ps.uplow

#subsetting your data by taxonomy
ps.rick <- subset_taxa(ps, Order=="Rickettsiales")
ps.rick

```
### Some standard filter/prune steps before you start the rest of your analyses:
```{r}
#get rid of 0 taxa
ps
ps_clean <- prune_taxa(taxa_sums(ps) > 0, ps)
any(sample_sums(ps_clean) == 0) #reports FALSE if no zero taxa
ps_clean
#get rid of taxa in less than 4 samples
ps_clean<- prune_taxa(taxa_sums(ps_clean) > 4, ps_clean)
ps_clean
#get rid of common contaminants
ps.filt<-subset_taxa(ps_clean, (Order!="Chloroplast") | is.na(Order)) 
ps.filt<-subset_taxa(ps_clean, (Family!="Mitochondria") | is.na(Family)) 
ps.filt
```



### Save your phyloseq object as a R data structure
This function allows you to save your phyloseq object to your working directory. This means that, in the future, you will only have to load your phyloseq object into R rather than 3 csv files. This saves a lot of time, keeps your folder cleaner, and makes working with others MUCH easier.

.rds files are r data structures. It saves one object in a way that it can then be loaded with saveRDS() later and R knows what it should look like.

```{r saving your phyloseq object to your working directory}
saveRDS(ps.filt,"examplePhyloseq.rds")
readRDS("examplePhyloseq.rds")
```



